spec/iot-js-spec.md

```diff
@@ -4,331 +4,476 @@ IoT Web API
 
-The stakeholders of this environment include:
+The devices support for resources can be extended via installable software modules called *applications*.
 
-  - the IoT *platform* provides mechanisms for developing solutions, exposed through the platform (e.g. IoTivity) API's and protocols.
+This API enables writing the applications that implement resources and business logic.
 
-  - vertical *solutions* using sensors, actuators, and contextual data, with solution *providers* managing one or more vertical solutions e.g. with local or cloud based services
+OIC high level Web API design notes
+-----------------------------------
+The API design starts minimal, solves the use cases relevant to JS developers, and gets expanded on demand.
 
-  - users with personal context.
+Code using this API is deployed to a device, which has one or more resources. In this version of the API it is assumed that the execution context of the code is separated for each device.
 
+Therefore the **API entry point** is an object that exposes the local device functionality, and can be requested with a ```require``` statement which invokes the constructor. When the object is garbage collected, the implementations should clean up native state.
 
-The basic expected business patterns are the following:
+**Device identification** is UUID. Each device has an associated address + port.
 
-  - a provider develops a solution, based on sensors and actuators,
+When a device is constructed, the implementation should announce its presence.
+Adding an event listener to the 'devicefound' event should turn on presence observing, i.e. the API implementation should make a request to watch presence notifications, and fire a 'devicefound' event when a presence notification is received.
 
-  - an installable software module (called *application*) is deployed in a gateway or dedicated solution box on the field, and
+**Resource identification** is URL path, relative to a given device.
 
-  - a cloud service with user accounts and profiles, remote management etc.
+On each device there are special resources, implementing device discovery, resource discovery, platform discovery, etc. Platform needs to be discoverable on a resource with a fixed URI ```/oic/p```, device on ```/oic/d``` and resources on ```/oic/res```. This API encapsulates these special resources and the hardcoded/fixed URIs by explicit function names and parameters.
 
-Multiple providers/solutions can share a gateway (a.k.a. OIC intermediary device);
-gateways' support for sensors can be extended via a software module as part of the solution.
+**Device discovery** uses endpoint discovery: multicast request "GET /oic/res" to "All CoAP nodes" (```224.0.1.187``` for IPv4 and ```FF0X::FD``` for IPv6, port 5683). The response lists devices and their resources (at least URI, resource type, interfaces, and media types). Since this is basically a resource discovery, it is merged with resource discovery.
 
```

Comment:
> **Device discovery** uses endpoint discovery: multicast request "GET /oic/res"

hm, we will use `oic/d` and not `oic/res?rt=oic.wk.d` for device discovery, right?